RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitPhase2WeightsProducer.cc

```diff
@@ -0,0 +1,109 @@
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+
+#include "DataFormats/EcalDigi/interface/EcalDigiCollections.h"
+#include "CondFormats/EcalObjects/interface/EcalCATIAGainRatios.h"
+#include "CondFormats/DataRecord/interface/EcalCATIAGainRatiosRcd.h"
+#include "DataFormats/EcalRecHit/interface/EcalRecHitCollections.h"
+#include "DataFormats/EcalRecHit/interface/EcalUncalibratedRecHit.h"
+#include "DataFormats/EcalDetId/interface/EBDetId.h"
+
+#include <TGraph.h>
+#include <TF1.h>
+
+class EcalUncalibRecHitPhase2WeightsProducer: public edm::stream::EDProducer<>{
+
+
+public:
+  explicit EcalUncalibRecHitPhase2WeightsProducer(const edm::ParameterSet& ps);//, edm::ConsumesCollector& c);
+  void produce(edm::Event& evt, const edm::EventSetup& es) override;
+
+private:
+  
+  float tRise_;
+  float tFall_;  
+  double gainRatios[2]={10,1};  
+  float gratio;
+
+  edm::EDGetTokenT<EBDigiCollectionPh2> ebDigiCollectionToken_;
+  std::string hitCollection_;
+
+};
+
+EcalUncalibRecHitPhase2WeightsProducer::EcalUncalibRecHitPhase2WeightsProducer(const edm::ParameterSet& ps){//, edm::ConsumesCollector& c){
+    
+  hitCollection_ = ps.getParameter<std::string>("EBhitCollection");
+  produces<EBUncalibratedRecHitCollection>(hitCollection_);
+  
+  tRise_ = ps.getParameter<double>("tRise");
+  tFall_ = ps.getParameter<double>("tFall");
+  
+  ebDigiCollectionToken_ = consumes<EBDigiCollectionPh2>(ps.getParameter<edm::InputTag>("BarrelDigis"));
+}
+
+void EcalUncalibRecHitPhase2WeightsProducer::produce(edm::Event& evt, const edm::EventSetup& es){
+  
+  // retrieve digis
+  edm::Handle<EBDigiCollectionPh2> pdigis_;
+  
+  evt.getByToken(ebDigiCollectionToken_, pdigis_);
+  
+  const EBDigiCollectionPh2* pdigis = nullptr;
+  pdigis = pdigis_.product();
+
+  // prepare output
+  auto ebUncalibRechits = std::make_unique<EBUncalibratedRecHitCollection>();
+  
+  for (auto itdg = pdigis->begin(); itdg != pdigis->end(); ++itdg) {
+    
+    EBDataFrame digi(*itdg);
+    EcalDataFrame_Ph2 dataFrame(*itdg);
+    DetId detId(digi.id());
+    
+    bool g1 = false; 
+    std::vector<float> timetrace;
+    std::vector<float> adctrace;
+    std::vector<double> weights = {-0.121016, -0.119899, -0.120923, -0.0848959, 0.261041, 0.509881, 0.373591, 0.134899, -0.0233605, -0.0913195, -0.112452, -0.118596, -0.120178, -0.12204, -0.121947, -0.122785};
```

Comment:
add a comment in the text saying "weights calculated averaging the full EB"

Reply:
could the weights be the same for the "pedestal" part?
How much would it change?
See the algebra

Reply:
Not necessary. The last four samples of the average pulse have slighty different ADC counts and so even the weights cannot be exactly equal.

Reply:
std::vector<double> weights = {-0.121016, -0.119899, -0.120923, -0.0848959, 0.261041, 0.509881, 0.373591, 0.134899, -0.0233605, -0.0913195, -0.112452, -0.118596, -0.120178, -0.12204, -0.121947, -0.122785};

becomes

std::vector<double> weights = ps.getParameter<vdouble>("weights");